yt_dlp/extractor/radiko.py

```diff
@@ -0,0 +1,228 @@
+
+        jscode = self._download_webpage(
+            'https://radiko.jp/apps/js/playerCommon.js', None,
+            note='Downloading player js code')
+        full_key = self._search_regex(
+            (r"RadikoJSPlayer\([^,]*,\s*(['\"])pc_html5\1,\s*(['\"])(?P<fullkey>[0-9a-f]+)\2,\s*{"),
+            jscode, 'full key', fatal=False, group='fullkey')
+
+        if full_key:
+            full_key = full_key.encode()
+        else:  # use full key ever known
+            full_key = b'bcd151073c03b352e1ef2fd66c32209da9ca0afa'
+
+        self._FULL_KEY = full_key
+        return full_key
+
+    def _find_program(self, video_id, station, cursor):
+        station_program = self._download_xml(
+            'https://radiko.jp/v3/program/station/weekly/%s.xml' % station, video_id,
+            note='Downloading radio program for %s station' % station)
+
+        prog = None
+        for p in station_program.findall('.//prog'):
+            ft_str, to_str = p.attrib['ft'], p.attrib['to']
+            ft = unified_timestamp(ft_str, False)
+            to = unified_timestamp(to_str, False)
+            if ft <= cursor and cursor < to:
+                prog = p
+                break
+        if not prog:
+            raise ExtractorError('Cannot identify radio program to download!')
+        assert ft, to
+        return prog, station_program, ft, ft_str, to_str
+
+    def _extract_formats(self, video_id, station, is_onair, ft, cursor, auth_token, area_id, query):
+        m3u8_playlist_data = self._download_xml(
+            'https://radiko.jp/v3/station/stream/pc_html5/%s.xml' % station, video_id,
+            note='Downloading m3u8 information')
+        m3u8_urls = m3u8_playlist_data.findall('.//url')
+
+        formats = []
+        found = set()
+        for url_tag in m3u8_urls:
+            pcu = url_tag.find('playlist_create_url')
+            url_attrib = url_tag.attrib
+            playlist_url = update_url_query(pcu.text, {
+                'station_id': station,
+                **query,
+                'l': '15',
+                'lsid': '77d0678df93a1034659c14d6fc89f018',
+                'type': 'b',
+            })
+            if playlist_url in found:
+                continue
+            else:
+                found.add(playlist_url)
+
+            time_to_skip = None if is_onair else cursor - ft
+
+            subformats = self._extract_m3u8_formats(
+                playlist_url, video_id, ext='m4a', entry_protocol='m3u8',
+                live=True, fatal=False, m3u8_id=None,
+                headers={
+                    'X-Radiko-AreaId': area_id,
+                    'X-Radiko-AuthToken': auth_token,
+                })
+            for sf in subformats:
+                domain = sf['format_id'] = compat_urllib_parse.urlparse(sf['url']).netloc
+                if re.match(r'^[cf]-radiko\.smartstream\.ne\.jp$', domain):
+                    sf['preference'] = 100 if is_onair else -100  # current radio stream
```

Comment:
this needs an explanation. the XML metadata reports that Radiko also serves the stream at c-radiko.smartstream.ne.jp and f-radiko.smartstream.ne.jp . (regex in line 126 matches these 2 domains)

these servers only serve current radio, not playback. so it has to be de-prioritized for RadikoIE (and not for RadikoRadioIE) before being taken in account for downloading. this is why L126-127 are needed.

L125 is to allow users to choose server by `-f` option, in case of problems occurred.